### PR TITLE
solution to review

### PR DESCRIPTION
**Pull request by ArturPoltser**

### CODE REVIEW

File: todo/models.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from django.db import models
+
+
+class Tag(models.Model):
+    name = models.CharField(max_length=100)
```
Comment (ArturPoltser) [inline todo/models.py:5]: Tag names should be unique, add `unique=True` param

File: todo/models.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    class Meta:
+        ordering = ["name"]
+
+    def __str__(self):
+        return f"{self.name}"
+
+
+class Task(models.Model):
+    content = models.CharField(blank=True, max_length=15)
```
Comment (ArturPoltser) [inline todo/models.py:15]: Hmm, `blank=True` might lead to tasks with no description - its incorrect. Also, `max_length=15` seems quite short for a task description... u should make it more usable, like 100 or 255

File: todo/models.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    deadline = models.DateTimeField()
```
Comment (ArturPoltser) [inline todo/models.py:17]: If deadlines are optional, consider `blank=True, null=True`. Also, how did u check if user add  `deadline` < than `created_at`?) 😄 Its important

File: todo/urls.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    path("task/create", TaskCreateView.as_view(), name="task-create"),
+    path("task/<int:pk>/update", TaskUpdateView.as_view(), name="task-update"),
+    path("task/<int:pk>/delete", TaskDeleteView.as_view(), name="task-delete"),
+
+    path("tags/", TagListView.as_view(), name="tag-list"),
+    path("tags/create", TagCreateView.as_view(), name="tag-create"),
+    path("tags/<int:pk>/update", TagUpdateView.as_view(), name="tag-update"),
+    path("tags/<int:pk>/delete", TagDeleteView.as_view(), name="tag-delete"),
+
+    path('chng_st/<int:pk>', TaskChangeStatusView.as_view(), name='task_change_status'),
```
Comment (ArturPoltser) [inline todo/urls.py:27]: its not explicit url name...better to use smts like `"task/<int:pk>/toggle-status"`. And don't forger to keep one style of quotes (double quotes prefered)

File: todo/views.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    template_name = "todo/task_list.html"
+    paginate_by = 3
+    queryset = Task.objects.prefetch_related("tags")
+
+
+class TaskCreateView(generic.CreateView):
+    model = Task
+    form_class = TaskForm
+    template_name = "todo/task_form.html"
+    success_url = "http://127.0.0.1:8000/task/create"
```
Comment (ArturPoltser) [inline todo/views.py:21]: Hardcoding full urls (with 127.0.0.1) is breaks in production or other environments. Use` reverse_lazy("todo:task-list")` or similar named routes instead

File: todo/views.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    context_object_name = "tag_list"
+    template_name = "todo/tag_list.html"
+    paginate_by = 3
+
+
+class TagCreateView(generic.CreateView):
+    model = Tag
+    form_class = TagForm
+    template_name = "todo/tag_form.html"
+    success_url = "http://127.0.0.1:8000/tags/create"
```
Comment (ArturPoltser) [inline todo/views.py:49]: same here ⬆️

File: todo/views.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+class TagDeleteView(generic.DeleteView):
+    model = Tag
+    fields = "__all__"
+    template_name = "todo/tag_confirm_delete.html"
+    success_url = reverse_lazy("todo:tag-list")
+
+
+class TaskChangeStatusView(View):
+    def post(self, request, pk):
+        task = Task.objects.get(id=pk)
```
Comment (ArturPoltser) [inline todo/views.py:68]: `get_object_or_404(Task, id=pk)` would be safer here to avoid 500 errors if the task doesn’t exist

File: todo/views.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+class TaskUpdateView(generic.UpdateView):
+    model = Task
+    form_class = TaskForm
+    template_name = "todo/task_form.html"
+    success_url = reverse_lazy("todo:task-list")
+
+
+class TaskDeleteView(generic.DeleteView):
+    model = Task
+    fields = "__all__"
```
Comment (ArturPoltser) [inline todo/views.py:33]: **fields** is not used in **DeleteView**, u can safely remove it

File: todo/forms.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        model = Tag
+        fields = "__all__"
+
+
+class TaskForm(forms.ModelForm):
+    tags = forms.ModelMultipleChoiceField(
+        queryset=Tag.objects.all(),
+        widget=forms.CheckboxSelectMultiple,
+        required=False
+    )
```
Comment (ArturPoltser) [inline todo/forms.py:18]: u need to provide here deadline validation for cheking that deadline is greater than now() 🙂  also good option here to add widget **DateTimeInput** for `deadline`

File: todo/views.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+class TaskChangeStatusView(View):
+    def post(self, request, pk):
+        task = Task.objects.get(id=pk)
+        task.is_done = not task.is_done
+        task.save()
+
+        return redirect(reverse_lazy('todo:task-list'), permanent=True)
+
+    def get(self, request):
+        return redirect(reverse_lazy('todo:task-list'), permanent=True)
```
Comment (ArturPoltser) [inline todo/views.py:75]: better to remove ` permanent=True` param (permanent=False by default), it returns 301 status, which is cached by browsers and may cause issues if the redirect target ever changes

**Review by ArturPoltser — COMMENTED** [review]
There is a couple of issues that needs to be fixed...also here a couple of advices for the future:
- create (or update) `.gitignore` file, u can fill it using this link [https://www.toptal.com/developers/gitignore](url) and selecting necessary parameters like Python, Django, etc. Because u push to git unnecessary files like .idea, db.sqlite3
- use one style of quotes (double quotes prefered)
- don't forget to write tests 🙂 it helps catch edge cases and unexpected behavior early
- you can also add JSON fixture with sample test data for better user experiance
